### PR TITLE
[Doc] Rewrite Fields Introduction

### PR DESCRIPTION
**Pull request by fzaninotto**
This is a central piece of react-admin, the doc should be crystal clear about props, usage, and recipes.

### CODE REVIEW

**Review by WiXSL — CHANGES_REQUESTED** [review]
**Already fixed**

[Datagrid.md line 110](https://github.com/marmelab/react-admin/blob/dae974b03b54ad92dea8d4ed06dd096490d5b0ac/docs/Datagrid.md?plain=1#L110)

```diff
- What's a Field component? Simply a component that reads the record (via `useGetRecord`) and renders a value. React-admin includes many Field componentds that you can use as children of `<Datagrid>` (`<TextField>`, `<NumberField>`, `<DateField>`, `<ReferenceField>`, and many more). Check [the Fields documentation](./Fields.md) for more information.
+ What's a Field component? Simply a component that reads the record (via `useRecordContext`) and renders a value. React-admin includes many Field components that you can use as children of `<Datagrid>` (`<TextField>`, `<NumberField>`, `<DateField>`, `<ReferenceField>`, and many more). Check [the Fields documentation](./Fields.md) for more information.

**Review by WiXSL — CHANGES_REQUESTED** [review]
[Datagrid.md line 116](https://github.com/marmelab/react-admin/blob/dae974b03b54ad92dea8d4ed06dd096490d5b0ac/docs/Datagrid.md?plain=1#L116)

```diff
- import * as React from "react";
+ import * as React from 'react';

**Review by WiXSL — CHANGES_REQUESTED** [review]
[Datagrid.md line 143](https://github.com/marmelab/react-admin/blob/dae974b03b54ad92dea8d4ed06dd096490d5b0ac/docs/Datagrid.md?plain=1#L143)

```diff
- Bulk action buttons are buttons that affect several records at once, like mass deletion for instance. In the `<Datagrid>` component, the bulk actions toolbar appears when a user ticks the checkboxes in the first column of the table. The user can then choose a button from the bulk actions toolbar. By default, all Datagrids have a single bulk action button, the bulk delete button. You can add other bulk action buttons by passing a custom element as the `bulkActionButtons` prop of the `<Datagrid>` component:
+ Bulk action buttons are buttons that affect several records at once, like mass deletion for instance. In the `<Datagrid>` component, the bulk actions toolbar appears when a user ticks the checkboxes in the first column of the table. The user can then choose a button from the bulk actions toolbar. By default, all datagrids have a single bulk action button, the bulk delete button. You can add other bulk action buttons by passing a custom element as the `bulkActionButtons` prop of the `<Datagrid>` component:

**Review by WiXSL — CHANGES_REQUESTED** [review]
**Already fixed**

[Datagrid.md line 530](https://github.com/marmelab/react-admin/blob/dae974b03b54ad92dea8d4ed06dd096490d5b0ac/docs/Datagrid.md?plain=1#L530)

```diff
- ```jsx
+ ```jsx
+ import { List, Datagrid, EditButton, BooleanField, DateField, TextField, useRecordContext } from 'react-admin';

**Comment by WiXSL** [comment]
I take the liberty to push directly some fixes, it was faster than point it them out in the review
Review still in progress...

File: docs/Fields.md [hunk 1]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
+| [`sortBy`](#sortby)           | Optional | `string`                       | `source` | When used in a `List`, specifies the actual `source` to be used for sorting when the user clicks the column header                                      |
+| [`sortByOrder`](#sortbyorder) | Optional | `ASC` &#124; `DESC`            | `ASC`    | When used in a `List`, specifies the sort order to be used for sorting when the user clicks the column header                                           |
+| [`className`](#classname)     | Optional | `string`                       | -        | A class name (usually generated by JSS) to customize the look and feel of the field element itself                                                      |
+| [`addLabel`](#addlabel)       | Optional | `boolean`                      | `false`  | Defines the visibility of the label when the field is used in `<SimpleForm>` or `<FormTab>`                                                             |
```
Comment (WiXSL) [inline docs/Fields.md:131]:
```suggestion
| [`addLabel`](#addlabel)       | Optional | `boolean`                      | `false`  | Defines the visibility of the label when the field is used in `<SimpleForm>` or `<TabbedForm>`                                                             |
```

File: docs/Fields.md [hunk 2]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
 
-const useStyles = makeStyles({
-    price: { color: 'purple' },
-});
+## `addLabel`
 
-const PriceField = props => {
-    const classes = useStyles();
-    return <TextField className={classes.price} {...props} />;
+Defines the visibility of the label when the field is used in a form (`<SimpleForm>` or `<FormTab>`). It’s true for all react-admin Field components.
```
Comment (WiXSL) [inline docs/Fields.md:138]:
```suggestion
Defines the visibility of the label when the field is used in a form (`<SimpleForm>` or `<TabbedForm>`). It’s true for all react-admin Field components.
```

File: docs/Fields.md [hunk 3]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
+
+If you create a custom Field component, and you want to use it in a form, set the `addLabel` defaultProp to `false` to have it labeled automatically.
```
Comment (WiXSL) [inline docs/Fields.md:140]:
```suggestion
If you create a custom Field component, and you want to use it in a form, set the `addLabel` defaultProp to `true` to have it labeled automatically.
```

File: docs/Fields.md [hunk 4]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
+
+```jsx
+const MyCustomMyField = props => (
+    /** ... **/
```
Comment (WiXSL) [inline docs/Fields.md:144]:
```suggestion
    /* ... */
```

File: docs/Fields.md [hunk 5]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
+);
+MyCustomMyField.defaultProps = {
+    addLabel: true,
 };
+```
+
+**Note**: This prop has no effect when rendering a field outside of a Form.
```
Comment (WiXSL) [inline docs/Fields.md:151]:
```suggestion
**Note**: This prop has no effect when rendering a field outside a Form.
```

File: docs/Fields.md [hunk 6]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
+// the label is automatically translated, so you can use translation identifiers
+<TextField source="author.name" label="ra.field.author" />
+// you can also use a React element
+<TextField source="author.name" label={<FieldTitle label="Author" />} />
 ```
 
-React-admin usually delegates the rendering of field components to material-ui components. Refer to the material-ui documentation to see the default styles for elements.
+**Tip**: If your admin has to support multiple languages, don't use the `label` prop, and put the localized labels in a dictionary instead. See the [Translation documentation](./Translation.md#translating-resource-and-field-names) for details.
+
+**Note**: This prop has no effect when rendering a field outside of a `<Datagrid>`, a `<SimpleShowLayout>`, a `<TabbedShowLayout>`, a `<SimpleForm>`, or a `<TabbedForm>`.
```
Comment (WiXSL) [inline docs/Fields.md:195]:
```suggestion
**Note**: This prop has no effect when rendering a field outside a `<Datagrid>`, a `<SimpleShowLayout>`, a `<TabbedShowLayout>`, a `<SimpleForm>`, or a `<TabbedForm>`.
```

File: docs/Fields.md [hunk 7]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
 
-You may want to customize the cell style inside a `Datagrid`. You can use the `cellClassName` for that:
+## `record`
+
+By default, fields use the `record` from the `RecordContext`. But you can override it by passing the `record` prop - e.g. if you're rendering a field outside of a `RecordContext`, or if you want to use a different record than the one in the context.
```
Comment (WiXSL) [inline docs/Fields.md:199]:
```suggestion
By default, fields use the `record` from the `RecordContext`. But you can override it by passing the `record` prop - e.g. if you're rendering a field outside a `RecordContext`, or if you want to use a different record than the one in the context.
```

File: docs/Fields.md [hunk 8]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
     </List>
 );
-
-// renders in the Datagrid as
-<td class="[class name generated by JSS]"><span>2</span></td>
 ```
-{% endraw %}
 
-You may want to override the field header (the `<th>` element in the `Datagrid`). In that case, use the `headerClassName` prop:
+**Note**: This prop has no effect when rendering a field outside of a `<Datagrid>`.
```
Comment (WiXSL) [inline docs/Fields.md:224]:
```suggestion
**Note**: This prop has no effect when rendering a field outside a `<Datagrid>`.
```

File: docs/Fields.md [hunk 9]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
+            </ReferenceField>
+        </Datagrid>
+    </List>
+);
+```
 
-const useStyles = makeStyles({
-    priceHeader: { fontWeight: 'bold' },
-});
+**Note**: This prop has no effect when rendering a field outside of a `<Datagrid>`.
```
Comment (WiXSL) [inline docs/Fields.md:245]:
```suggestion
**Note**: This prop has no effect when rendering a field outside a `<Datagrid>`.
```

File: docs/Fields.md [hunk 10]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
         </Datagrid>
     </List>
 );
-// renders in the table header as
-<th class="[class name generated by JSS]"><button>Price</button></th>
 ```
-{% endraw %}
 
-Finally, sometimes, you just want to right align the text of a cell. Use the `textAlign` prop, which accepts either `left` or `right`:
+**Note**: This prop has no effect when rendering a field outside of a `<Datagrid>`.
```
Comment (WiXSL) [inline docs/Fields.md:264]:
```suggestion
**Note**: This prop has no effect when rendering a field outside a `<Datagrid>`.
```

File: docs/Fields.md [hunk 11]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
+import { List, Datagrid, WrapperField, TextField } from 'react-admin';
+
+const UserList = () => (
+   <List>
+       <Datagrid>
+            <ImageField source="avatar" sx={{ my: -2 }}/>
+            <TextField source="username" sx={{ color: 'lightgrey' }} />
+            <TextField source="email" sx={{ textOverflow: 'ellipsis" }} />
+      </Datagrid>
+  </List>
```
Comment (WiXSL) [inline docs/Fields.md:289]:
```suggestion
    <List>
        <Datagrid>
            <ImageField source="avatar" sx={{ my: -2 }}/>
            <TextField source="username" sx={{ color: 'lightgrey' }} />
            <TextField source="email" sx={{ textOverflow: 'ellipsis' }} />
        </Datagrid>
    </List>
```

File: docs/Fields.md [hunk 12]
```diff
@@ -359,122 +529,73 @@ const MyShowLayout = () => (
...
+            <TextField source="first_name" />
+            <TextField source="last_name" />
+            <ConditionalEmailField />
+        </SimpleShowLayout>
+    </Show>
 );
 ```
 
-The `<ShowController>` fetches the `record` from the data provider, and passes it to its child function when received (among the `controllerProps`). That means the following code:
+This `<ConditionalEmailField>` is properly hidden when `hasEmail` is `false`. But the label for the field never renders. And if you add a `label` default prop, `SimpleShowLayout` layout will render the label regardless of the `hasEmail` value...
```
Comment (WiXSL) [inline docs/Fields.md:557]:
```suggestion
This `<ConditionalEmailField>` is properly hidden when `hasEmail` is `false`. But the label for the field never renders. And if you add a `label` default prop, `SimpleShowLayout` layout will render the label regardless of the `hasEmail` value.
```

File: docs/Fields.md [hunk 13]
```diff
@@ -359,122 +529,73 @@ const MyShowLayout = () => (
...
-    <Show {...props}>
+```jsx
+const UserShow = () => (
+    <Show>
         <SimpleShowLayout>
-            <TextField source="username" />
-            <TextField source="email" />
+            <TextField source="first_name" />
+            <TextField source="last_name" />
+            {/** where can we get the record? **/}
```
Comment (WiXSL) [inline docs/Fields.md:567]:
```suggestion
            {/* Where can we get the record? */}
```

File: docs/Fields.md [hunk 14]
```diff
@@ -359,122 +529,73 @@ const MyShowLayout = () => (
...
+const UserShowLayout = () => {
+    const record = useRecordContext();
+    if (!record) return null;
+    return (
+        <SimpleShowLayout>
+            <TextField source="first_name" />
+            <TextField source="last_name" />
+            {record.hasEmail && <EmailField source="email" />}
+        </SimpleShowLayout>
+    ):
```
Comment (WiXSL) [inline docs/Fields.md:592]:
```suggestion
    );
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Comment by WiXSL** [comment]
Review finished

File: docs/Fields.md [hunk 15]
```diff
@@ -58,219 +44,380 @@ const PostShow = ({ id }) => {
...
+| [`sortable`](#sortable)       | Optional | `boolean`                      | `true`   | When used in a `List`, should the list be sortable using the `source` attribute? Setting it to `false` disables the click handler on the column header. |
+| [`sortBy`](#sortby)           | Optional | `string`                       | `source` | When used in a `List`, specifies the actual `source` to be used for sorting when the user clicks the column header                                      |
+| [`sortByOrder`](#sortbyorder) | Optional | `ASC` &#124; `DESC`            | `ASC`    | When used in a `List`, specifies the sort order to be used for sorting when the user clicks the column header                                           |
+| [`className`](#classname)     | Optional | `string`                       | -        | A class name (usually generated by JSS) to customize the look and feel of the field element itself                                                      |
```
Comment (djhi) [inline docs/Fields.md:130]: I think we're missing `headerClassName` and `cellClassName`. Did we remove them already?

**Review by djhi — COMMENTED** [review]
Nice!

File: docs/Fields.md (hunk 15 above)
Comment (fzaninotto) [reply docs/Fields.md:130]: We didn't remove them, but I chose not to document them anymore. With the `sx` prop, they are no longer the best way to style a column. It's documented in Datagrid.md

**Review by djhi — APPROVED** [review]

File: docs/Fields.md (hunk 15 above)
Comment (djhi) [reply docs/Fields.md:130]: I missed the part referencing a field :+1:

**Pull request merged**